_includes/source_code/code/24-PDP/souvlakia/souvlakia_nlogn_segtree.cc

```diff
@@ -0,0 +1,187 @@
+#ifdef CONTEST
+    #define NDEBUG    //disable assert() checks
+#endif
+
+#include <cstdio>
+#include <algorithm>
+#include <climits>
+#include <cstdint>
+#include <cassert>
+#include <vector>
+#include <list>
+#include <cstring>
+#include <cmath>
+#include <set>
+
+using namespace std;
+
+const int32_t 
+    MAXN = int32_t(1e5), //max shops
+    MAXQ = int32_t(5e4), //max queries
+    INF  = INT32_MAX; //INF for dijkstra
+
+int32_t    N,M,Q,C[3];
+
+struct shop {
+#define X  d[0]
+#define Y  d[1]
+#define Z  d[2]
+    int32_t d[3];//The 3 distances (X,Y,Z) from {A,B,C}
+    list<int32_t> queryid;//query ids that has to answer shop[i]
+            //use list in case there are more than one queries per shop
+    shop(){ X=Y=Z=INF; }//help dijsktra initial values
+    bool operator <(const shop& b) const { 
+        return X < b.X;
+    }
+} S[MAXN+1];
+
+void dijsktra(int32_t src,int dupd,vector<vector<pair<int32_t,int32_t>>>& edge){
+    //src is one of {A,B,C}
+    //dupd in [0,3). Update Shop[*].d[dupd] distance from src
+    set<pair<int32_t,int32_t>> DS;//<distance,shopid>
+    
+    DS.insert({0,src});
+    S[src].d[dupd] = 0;
+    
+    while(!DS.empty()){//explore
+        auto x = *(DS.begin());
+        DS.erase(DS.begin());
+        int32_t daddy = x.second;
+        for(auto y : edge[daddy]){//expand
+            int32_t child = y.first, dist = y.second;
+            if(S[daddy].d[dupd]!=INF && S[child].d[dupd] > S[daddy].d[dupd] + dist){
+                if(S[child].d[dupd]!=INF)
+                    DS.erase(DS.find({S[child].d[dupd],child}));
+                S[child].d[dupd] = S[daddy].d[dupd] + dist;
+                DS.insert({S[child].d[dupd],child});
+            }
+        }
+    }
+}
+
+inline int32_t read_fast(){//fast IO is required for this problem due to huge io
+    int32_t x = 0;
+    char c;
+    while((c=getchar_unlocked())<'0');//skip blanks, CR, LF
+    do
+        x = x*10 + (c-'0');
+    while((c=getchar_unlocked())>='0');
+    return x;
+}
+
+inline void write_fast(bool f){
+    const char *s = f ? "YES\n":"NO\n";
+    while(*s)
+        putchar_unlocked(*s++);
+}
+
+//RMQ with segment tree code starts here
+int32_t YMAX = 1;    //max value of Y
+int32_t *Zrmq;//segment tree (RangeMinQuery)for Z
+void sinit(){//allocate mem for segment tree and initialize it to (almost) infinity
+    for(int32_t i=1;i<=N;++i)
+        YMAX = max(YMAX, S[i].Y);
+    size_t Zrmqsize = 1 << size_t(ceil(log2(YMAX+1))+1);
+    Zrmq = new int32_t[Zrmqsize];
+    memset(Zrmq,0x7f,sizeof(*Zrmq) * Zrmqsize);
```

Comment:
Έχω προσθέσει ένα footnote να εξηγεί την επιλογή της δεύτερης παραμέτρου της memset (δεν έχει γίνει ακόμα push):

[^13]: Για το segment tree χρησιμοποιήθηκε ένας απλός πίνακας προσημασμένων ακεραίων 32bit. Θέλουμε να τον αρχικοποιήσουμε σε τιμή $$+\infty$$ ή έστω κάποια πολύ μεγάλη τιμή όπως το $$\mathit{INT\_MAX}$$. Μια λύση για να το καταφέρουμε είναι με ένα for βρόγχο. Άλλη λύση είναι με τη χρήση της memset. Η memset όμως αποθηκεύει μια τιμή σε κάθε byte του πίνακα και όχι σε κάθε ακέραιο χωριστά. Το μέγιστο byte που θα μπορούσαμε να περάσουμε στη memset είναι το $$255$$ (ή 0xFF), όμως αυτό θα θέσει (ανάψει) και το MSB (πιο σημαντικό bit) του ακεραίου, κάνοντας τον αρνητικό, πράγμα που δεν θέλουμε. Το byte που περνάμε στη memset έχει το MSB σβηστό και είναι το $$127$$ (0x7F) και οι ακέραιοι μας θα πάρουν την τιμή $$‭2139062143‬$$ (0x7F7F7F7F)που είναι πολύ κοντά στο $$\mathit{INT\_MAX}$$. Άλλη λύση είναι να χρησιμοποιήσουμε std::vector(αριθμός_στοιχείων, αρχική_τιμή).
![φοοτ13](https://user-images.githubusercontent.com/62041257/80324070-8be76d00-8837-11ea-800c-b3d7465efa93.jpg)